File: LoadingAndSaving/LoadingAndSavingHTML/Word2Help/VisualBasic/Topic.vb
'////////////////////////////////////////////////////////////////////////
' Copyright 2001-2014 Aspose Pty Ltd. All Rights Reserved.
'
' This file is part of Aspose.Words. The source code in this file
' is only intended as a supplement to the documentation, and is provided
' "as is", without warranty of any kind, either expressed or implied.
'////////////////////////////////////////////////////////////////////////

Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Text
Imports Aspose.Words
Imports Aspose.Words.Fields
Imports Aspose.Words.Saving

Namespace Word2HelpExample
	''' <summary>
	''' Represents a single topic that will be written as an HTML file.
	''' </summary>
	Public Class Topic
		''' <summary>
		''' Creates a topic.
		''' </summary>
		Public Sub New(ByVal section As Section, ByVal fixUrl As String)
			mTopicDoc = New Document()
			mTopicDoc.AppendChild(mTopicDoc.ImportNode(section, True, ImportFormatMode.KeepSourceFormatting))
			mTopicDoc.FirstSection.Remove()

			Dim headingPara As Paragraph = CType(mTopicDoc.FirstSection.Body.FirstChild, Paragraph)
			If headingPara Is Nothing Then
				ThrowTopicException("The section does not start with a paragraph.", section)
			End If

			mHeadingLevel = headingPara.ParagraphFormat.StyleIdentifier - StyleIdentifier.Heading1
			If (mHeadingLevel < 0) OrElse (mHeadingLevel > 8) Then
				ThrowTopicException("This topic does not start with a heading style paragraph.", section)
			End If

			mTitle = headingPara.GetText().Trim()
			If mTitle = "" Then
				ThrowTopicException("This topic heading does not have text.", section)
			End If

			' We actually remove the heading paragraph because <h1> will be output in the banner.
			headingPara.Remove()

			mTopicDoc.BuiltInDocumentProperties.Title = mTitle

			FixHyperlinks(section.Document, fixUrl)
		End Sub

		Private Shared Sub ThrowTopicException(ByVal message As String, ByVal section As Section)
			Throw New Exception(message & " Section text: " & section.Body.ToString(SaveFormat.Text).Substring(0, 50))
		End Sub

		Private Sub FixHyperlinks(ByVal originalDoc As DocumentBase, ByVal fixUrl As String)
			If fixUrl.EndsWith("/") Then
				fixUrl = fixUrl.Substring(0, fixUrl.Length - 1)
			End If

			Dim fieldStarts As NodeCollection = mTopicDoc.GetChildNodes(NodeType.FieldStart, True)
			For Each fieldStart As FieldStart In fieldStarts
				If fieldStart.FieldType <> FieldType.FieldHyperlink Then
					Continue For
				End If

				Dim hyperlink As New Hyperlink(fieldStart)
				If hyperlink.IsLocal Then
					' We use "Hyperlink to a place in this document" feature of Microsoft Word
					' to create local hyperlinks between topics within the same doc file.
					' It causes MS Word to auto generate the bookmark name.
					Dim bmkName As String = hyperlink.Target

					' But we have to follow the bookmark to get the text of the topic heading paragraph
					' in order to be able to build the proper filename of the topic file.
					Dim bmk As Bookmark = originalDoc.Range.Bookmarks(bmkName)

					If bmk Is Nothing Then
						Throw New Exception(String.Format("Found a link to a bookmark, but cannot locate the bookmark. Name:'{0}'.", bmkName))
					End If

					Dim para As Paragraph = CType(bmk.BookmarkStart.ParentNode, Paragraph)
					Dim topicName As String = para.GetText().Trim()

					hyperlink.Target = HeadingToFileName(topicName) & ".html"
					hyperlink.IsLocal = False
				Else
					' We "fix" URL like this:
					' http://www.aspose.com/Products/Aspose.Words/Api/Aspose.Words.Body.html
					' by changing them into this:
					' Aspose.Words.Body.html
					If hyperlink.Target.StartsWith(fixUrl) AndAlso (hyperlink.Target.Length > (fixUrl.Length + 1)) Then
						hyperlink.Target = hyperlink.Target.Substring(fixUrl.Length + 1)
					End If
				End If
			Next fieldStart
		End Sub

		Public Sub WriteHtml(ByVal htmlHeader As String, ByVal htmlBanner As String, ByVal htmlFooter As String, ByVal outDir As String)
			Dim fileName As String = Path.Combine(outDir, Me.FileName)

			Dim saveOptions As New HtmlSaveOptions()
			saveOptions.PrettyFormat = True
			' This is to allow headings to appear to the left of main text.
			saveOptions.AllowNegativeLeftIndent = True
			' Disable headers and footers.
			saveOptions.ExportHeadersFootersMode = ExportHeadersFootersMode.None

			' Export the document to HTML.
			mTopicDoc.Save(fileName, saveOptions)

			' We need to modify the HTML string, read HTML back.
			Dim html As String
			Using reader As New StreamReader(fileName)
				html = reader.ReadToEnd()
			End Using

			' Builds the HTML <head> element.
			Dim header As String = RegularExpressions.HtmlTitle.Replace(htmlHeader, mTitle, 1)

			' Applies the new <head> element instead of the original one.
			html = RegularExpressions.HtmlHead.Replace(html, header, 1)
			html = RegularExpressions.HtmlBodyDivStart.Replace(html, " id=""nstext""", 1)

			Dim banner As String = htmlBanner.Replace("###TOPIC_NAME###", mTitle)

			' Add the standard banner.
			html = html.Replace("<body>", "<body>" & banner)

			' Add the standard footer.
			html = html.Replace("</body>", htmlFooter & "</body>")

			Using writer As New StreamWriter(fileName)
				writer.Write(html)
			End Using
		End Sub

		''' <summary>
		''' Removes various characters from the heading to form a file name that does not require escaping.
		''' </summary>
		Private Shared Function HeadingToFileName(ByVal heading As String) As String
			Dim b As New StringBuilder()
			For Each c As Char In heading
				If Char.IsLetterOrDigit(c) Then
					b.Append(c)
				End If
			Next c

			Return b.ToString()
		End Function

		Public ReadOnly Property Document() As Document
			Get
				Return mTopicDoc
			End Get
		End Property

		''' <summary>
		''' Gets the name of the topic html file without path.
		''' </summary>
		Public ReadOnly Property FileName() As String
			Get
				Return HeadingToFileName(mTitle) & ".html"
			End Get
		End Property

		Public ReadOnly Property Title() As String
			Get
				Return mTitle
			End Get
		End Property

		Public ReadOnly Property HeadingLevel() As Integer
			Get
				Return mHeadingLevel
			End Get
		End Property

		''' <summary>
		''' Returns true if the topic has no text (the heading paragraph has already been removed from the topic).
		''' </summary>
		Public ReadOnly Property IsHeadingOnly() As Boolean
			Get
				Dim body As Body = mTopicDoc.FirstSection.Body
				Return (body.FirstParagraph Is Nothing)
			End Get
		End Property

		Private ReadOnly mTopicDoc As Document
		Private ReadOnly mTitle As String
		Private ReadOnly mHeadingLevel As Integer
	End Class
End Namespace

File: Examples/LoadingAndSaving/LoadingAndSavingHTML/Word2Help/CSharp/Topic.cs
//////////////////////////////////////////////////////////////////////////
// Copyright 2001-2014 Aspose Pty Ltd. All Rights Reserved.
//
// This file is part of Aspose.Words. The source code in this file
// is only intended as a supplement to the documentation, and is provided
// "as is", without warranty of any kind, either expressed or implied.
//////////////////////////////////////////////////////////////////////////
using System;
using System.IO;
using System.Text;
using Aspose.Words;
using Aspose.Words.Fields;
using Aspose.Words.Saving;

namespace Word2HelpExample
{
    /// <summary>
    /// Represents a single topic that will be written as an HTML file.
    /// </summary>
    public class Topic
    {
        /// <summary>
        /// Creates a topic.
        /// </summary>
        public Topic(Section section, string fixUrl)
        {
            mTopicDoc = new Document();
            mTopicDoc.AppendChild(mTopicDoc.ImportNode(section, true, ImportFormatMode.KeepSourceFormatting));
            mTopicDoc.FirstSection.Remove();

            Paragraph headingPara = (Paragraph)mTopicDoc.FirstSection.Body.FirstChild;
            if (headingPara == null)
                ThrowTopicException("The section does not start with a paragraph.", section);

            mHeadingLevel = headingPara.ParagraphFormat.StyleIdentifier - StyleIdentifier.Heading1;
            if ((mHeadingLevel < 0) || (mHeadingLevel > 8))
                ThrowTopicException("This topic does not start with a heading style paragraph.", section);
            
            mTitle = headingPara.GetText().Trim();
            if (mTitle == "")
                ThrowTopicException("This topic heading does not have text.", section);

            // We actually remove the heading paragraph because <h1> will be output in the banner.
            headingPara.Remove();

            mTopicDoc.BuiltInDocumentProperties.Title = mTitle;

            FixHyperlinks(section.Document, fixUrl);
        }

        private static void ThrowTopicException(string message, Section section)
        {
            throw new Exception(message + " Section text: " + section.Body.ToString(SaveFormat.Text).Substring(0, 50));
        }

        private void FixHyperlinks(DocumentBase originalDoc, string fixUrl)
        {
            if (fixUrl.EndsWith("/"))
                fixUrl = fixUrl.Substring(0, fixUrl.Length - 1);

            NodeCollection fieldStarts = mTopicDoc.GetChildNodes(NodeType.FieldStart, true);
            foreach (FieldStart fieldStart in fieldStarts)
            {
                if (fieldStart.FieldType != FieldType.FieldHyperlink)
                    continue;

                Hyperlink hyperlink = new Hyperlink(fieldStart);
                if (hyperlink.IsLocal)
                {
                    // We use "Hyperlink to a place in this document" feature of Microsoft Word
                    // to create local hyperlinks between topics within the same doc file.
                    // It causes MS Word to auto generate the bookmark name.
                    string bmkName = hyperlink.Target;

                    // But we have to follow the bookmark to get the text of the topic heading paragraph
                    // in order to be able to build the proper filename of the topic file.
                    Bookmark bmk = originalDoc.Range.Bookmarks[bmkName];

                    if (bmk == null)
                        throw new Exception(string.Format("Found a link to a bookmark, but cannot locate the bookmark. Name:'{0}'.", bmkName));

                    Paragraph para = (Paragraph)bmk.BookmarkStart.ParentNode;
                    string topicName = para.GetText().Trim();

                    hyperlink.Target = HeadingToFileName(topicName) + ".html";
                    hyperlink.IsLocal = false;
                }
                else
                {
                    // We "fix" URL like this:
                    // http://www.aspose.com/Products/Aspose.Words/Api/Aspose.Words.Body.html
                    // by changing them into this:
                    // Aspose.Words.Body.html
                    if (hyperlink.Target.StartsWith(fixUrl) &&
                        (hyperlink.Target.Length > (fixUrl.Length + 1)))
                    {
                        hyperlink.Target = hyperlink.Target.Substring(fixUrl.Length + 1);
                    }
                }
            }
        }

        public void WriteHtml(string htmlHeader, string htmlBanner, string htmlFooter, string outDir)
        {
            string fileName = Path.Combine(outDir, FileName);

            HtmlSaveOptions saveOptions = new HtmlSaveOptions();
            saveOptions.PrettyFormat = true;
            // This is to allow headings to appear to the left of main text.
            saveOptions.AllowNegativeLeftIndent = true;
            // Disable headers and footers.
            saveOptions.ExportHeadersFootersMode = ExportHeadersFootersMode.None;

            // Export the document to HTML.
            mTopicDoc.Save(fileName, saveOptions);

            // We need to modify the HTML string, read HTML back.
            string html;
            using (StreamReader reader = new StreamReader(fileName))
                html = reader.ReadToEnd();

            // Builds the HTML <head> element.
            string header = RegularExpressions.HtmlTitle.Replace(htmlHeader, mTitle, 1);
            
            // Applies the new <head> element instead of the original one.
            html = RegularExpressions.HtmlHead.Replace(html, header, 1);
            html = RegularExpressions.HtmlBodyDivStart.Replace(html, @" id=""nstext""", 1);

            string banner = htmlBanner.Replace("###TOPIC_NAME###", mTitle);
            
            // Add the standard banner.
            html = html.Replace("<body>", "<body>" + banner);
            
            // Add the standard footer.
            html = html.Replace("</body>", htmlFooter + "</body>");

            using (StreamWriter writer = new StreamWriter(fileName))
                writer.Write(html);
        }

        /// <summary>
        /// Removes various characters from the heading to form a file name that does not require escaping.
        /// </summary>
        private static string HeadingToFileName(string heading)
        {
            StringBuilder b = new StringBuilder();
            foreach (char c in heading)
            {
                if (Char.IsLetterOrDigit(c))
                    b.Append(c);
            }

            return b.ToString();
        }

        public Document Document
        {
            get { return mTopicDoc; }
        }

        /// <summary>
        /// Gets the name of the topic html file without path.
        /// </summary>
        public string FileName
        {
            get { return HeadingToFileName(mTitle) + ".html"; }
        }

        public string Title
        {
            get { return mTitle; }
        }

        public int HeadingLevel
        {
            get { return mHeadingLevel; }
        }

        /// <summary>
        /// Returns true if the topic has no text (the heading paragraph has already been removed from the topic).
        /// </summary>
        public bool IsHeadingOnly
        {
            get
            {
                Body body = mTopicDoc.FirstSection.Body;
                return (body.FirstParagraph == null);
            }
        }

        private readonly Document mTopicDoc;
        private readonly string mTitle;
        private readonly int mHeadingLevel;
    }
}